README.md
# Kubernetes API Browser (KAB)

## 0. Introduction to the KAB tool

The Kubernetes API is a huge collection of endpoints that serve different
resource types. All users interact with the API server in various ways.
However, the current API specification and its documentation has some
limitations that are difficult to overcome.

1. The API specification is generated from the comments in the Go source
   code which may be imprecise and limited in their capability to leverage
   the OpenAPI constructs.
1. The API specification generator relys on a limited set of directives in
   the Go source comments for parsing, so there is not much room for the tool
   to close the gap.
1. There is no suitable browser for the Kubernetes OpenAPI specification.

The Kubernetes API Browser (KAB) tool is designed to resolve the above
problems by providing an interactive browser backed by manually optimized
specification contents.

The usage of this tool is simple:

```
docker run -p 8000:8000 quay.io/tengqm/kab
```

Then point your browser to 'http://localhost:8000' and start exploring.

## 1. Problems of Generated OpenAPI Specification

The Kubernetes API specification is generated from Go source code comments,
which are primarily targeted at Kubernetes developers instead of its users.
The situation is worsened by the lack of a global guideline to code comments.

While the developer community is to some extent aware of these inaccuracies or
mistakes, it is still impractical to completely avoid such problems.

### 1.1 Comments are for developers

The shared constructs also generates a lot of redundant (and sometimes
meaningless) text in the API spec. For example, the `apiVersion` field for an
API resource is currently documented as the following:

```json
{
  "apiVersion": {
     "description": "APIVersion defines the versioned schema of this representation of an object. Servers should convert recognized schemas to the latest internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources",
     "type": "string"
  },
}
```

This document is generated from the shared comment for the `APIVersion` Go
field, but it is not targeted at users. One of the common issues of the API
specification is that the fields of types are often times referenced by their
Go member names (e.g. `APIVersion`) rather than the API field names presented
to user (e.g. `apiVersion`).  When these names are exposed to users, they may
get confused what the document is referring to. 

As for the `description` field, the content is not targeting end users either.
A user won't care how the server converts the schema to internal values and
the pointer to the API convention design document is irrelevant to Kubernetes
users.

In KAB, fields like this are revised on a per-package, per-type basis. For
example, for the definition (resource) `io.k8s.api.storage.v1.CSIDriver`, the
`apiVersion` field is documented as:

```json
{
  "apiVersion": {
    "description": "The version of the schema for the object representation",
    "enum": [
      "storage.k8s.io/v1"
    ],
    "type": "string"
  }
}
```

### 1.2 Markdown comment assumption fails everywhere

There is an implicit assumption that the comments are somehow written in
Markdown syntax for document typesetting. Consequently, the correct generation
of documentation relies on proper usage of Markdown syntax in the comment,
which are difficult to enforce across the community. A missing empty line may
lead to an incorrectly rendered list. For example, the following comment looks
good enough for Go developers until we want to extract the comment as Markdown
text. To make a Markdown parser happy, we have to add an empty line before the
list, or else the rendered output will be unreadable.

```Go
  // addressType specifies the type of address carried by this EndpointSlice.
  // All addresses in this slice must be the same type. This field is
  // immutable after creation. The following address types are currently
  // supported:
  // * IPv4: Represents an IPv4 Address.
  // * IPv6: Represents an IPv6 Address.
  // * FQDN: Represents a Fully Qualified Domain Name.
  AddressType AddressType
```

### 1.3 Limitations 

Generating documentation from source code helps improve the consistency
between the implementation and the documentation, however, for API
specification, the amount of constraints or conditions may eventually prove
that we have to maintain the specification separately rather than "document"
all the details using language extensions.

The API specification generator today relies on many source code level
annotations for documenting the field level constraints. The following are
some examples:

- `+optional`: This used to indicate a field is not mandatory. There are many
  places this annotation is incorrectly used. Considering that we are not
  differentiating definitions used for different requests, such a problem is
  aggravated. What's more, this annotation sometimes conflicts with the
  comment. For example, one cannot tell if the `IQN` field is required or not
  based on the following snippet:

  ```Go
  type ISCSIVolumeSource struct {
    // Required: target iSCSI Qualilfied Name
    // +optional
    IQN string
    ...
  }
  ```

- `+unionDiscriminator`: This is used to mark a field that is used to
  differentiate two different structs.

- `+featureGate=...`: This is used to relate a field to the name of the
  feature gate that signifies the stage for the field/feature.

There are ongoing efforts to add more annotations. There are many constraints
that can be covered only by informal and inconsistent Go comments rather than
any annotations. The following are a few examples:

- The value range of an integer field. This is supported in OpenAPI
  specification using the `minimum` and/or `maximum` attributes, but only
  informally and inconsistently captured in the Go source comments.

  In KAB, such constraints are manually added to the specification based on
  the validation logic. For example:

  ```yaml
  # io.k8s.api.core.v1.PodSpec.CREATE
  properties:
    activeDeadlineSeconds:
      description: "Duration in seconds the Pod may be ..."
      format: int64
      minimum: 1
      type: integer
  ```

- The length of a string field. For string that is required, most of the time
  the semantic is that the string must be provided and it cannot be empty.
  This constraint is supported in OpenAPI specification using the `minLength`
  attribute, but it is can only documented in the Go source comments.

  In KAB, such constraints are manually added to the specification based on
  the validation logic. For example:

  ```yaml
  # io.k8s.api.core.v1.Container
  properties:
    name: "Name of the container. Each container in ..."
    minLength: 1
    type: string
  ```


- The size of an array. For array/list type of fields, there may be constraint
  that the array cannot be empty. Such constraints are not formally captured
  or documented.

  In KAB, this constraint is added based on the validation logic. For example:

  ```yaml
  # io.k8s.api.core.v1.Pod.CREATE
  properties:
    containers:
      description: "List of containers in the Pod. Containers ..."
      items:
        $ref: #/definitions/io.k8s.api.core.v1.Container
      minItems: 1
      type: array
      x-kubernetes-patch-merge-key: name
      x-kubernetes-patch-strategy: merge
  ```

- The default value. Default values are not formally documented although
  supported in OpenAPI specification using the `default` attribute. 

- The stages of fields are not formally documented. For example, an Alpha
  field maybe documented as "ALPHA", "Alpha" or "alpha" in the comments.
  Deprecated fields or types are not formally documented either. The related
  comment may be "DEPRECATED", "deprecated" or "Deprecated". The only way to
  find out which fields are going away is to grep the whole specification and
  manually parse the comments.

In KAB, the field level constraints are carefully amended based on the builtin
validation logic. OpenAPI keywords are added where appropriate. When there is
no related OpenAPI attribute keyword, the field/type comment is amended to
include these details.


## 2. Problems from the generator

In addition to the problems found in the code comments, there are issues
related to the limitations of the generator.

### 2.1 No differentiation among `CREATE`, `UPDATE` and `GET`

The same resource may have different requirements to its fields. For a `CREATE`
request, some fields are mandatory and other fields are optional. For an
`UPDATE` request, the set of mandatory fields might be different from that of
a `CREATE` request. There are fields that cannot be updated once specified.
For example, you are not supposed to specify the `.status` field when creating
an object, but when you are updating an existing object, you are permitted to
mutate its `.status` field. For another example, you may have to specify the
`.metadata.name` field when creating an object, but you don't have to provide
this field when updating it.

When you are reading an object or a list of objects, you will likely get all
the fields back except for those "write-only" ones. This effectively means
that for a `GET` request, all fields of a given type may be mandatory. 

### 2.2 Comments for embedded structs 

There is no inheritance in the Go language. Nested structs is the preferred
way to capture common attributes among different groups of objects. This
language design presents challenges for documentation generated from the
code comments. For example, the `ObjectMeta` struct shared by many API
resources contains the `Name` field which is presented to users as the
`.metadata.name` field. Different resource types may have different
requirements to their names, but there is only one place to specify the
comment for this field. It is difficult to highlight that, for example, the
name for a `CronJob` object must be less than 53 characters.

### 2.3 Documents are too generic to be useful

Take another look at the `apiVersion` document above, the document correctly
specifies that the type of this field is a `string`, but it fails to indicate
what is the expected value of the string. For almost all API types taht are
documented today, this `apiVersion` field has a fixed value. For example, for
`io.k8s.api.core.v1.ConfigMap`, this has to be "`v1`"; for
`io.k8s.api.storage.v1.CSIDriver`, this has to be "`storage.k8s.io/v1`"; for
`io.k8s.api.rbac.v1.ClusterRole`, this has to be
"`rbac.authorization.k8s.v1`".

In KAB, fields like these are treated as having a constant string value in the
`enum` field. The definition of the `Pod` resource type then begins with the
following snippet:

```
{
  "apiVersion": {
    "description": "The version of the schema for the object representation",
    "enum": [
      "v1"
    ],
    "type": "string"
  },
  "kind": {
    "description": "A string value representing the REST resource",
    "enum": [
      "Pod"
    ],
    "type": "string"
  },
  ...
}
```

### 2.4 Enumeration for fields 

For example, the `imagePullPolicy` of the `v1/Container` type can be one of
"`Always`", "`IfNotPresent`", and "`Never`". The value set is defined as three
standalone string constants in the source code due to lack of `enum`
constructs in the Go language. The consequence is that the API specification
generator cannot safely deduce the safe set of valid values by analyzing the
source code. The best thing it can do is to specify that the `imagePullPolicy`
field is a string. Such an obvious enumeration can documented only as source
code comments. However, the developers still fail to do so in many cases,
because in the source code, everything looks just fine.

In KAB, fields like this are documented explicitly using `enum`s, e.g.

```yaml
  # ...
  imagePullPolicy:
    description: "Image pull policy. Defaults to `Always` if the `:latest` tag is specified."
    enum:
      - Always
      - IfNotPresent
      - Never
    type: string
```

### 2.5 Duplicated operation parameters

Many operations in Kubernetes API share the same collection of parameters.
In the OpenAPI specification, shared parameters can be captured in a dedicated
section (`parameters`). Because the Kubernetes API specification is generated
on a per-package, per-type basis, such global optimization cannot be applied.
Instead, what we get is a huge API specification containing many ducplicated
contents regarding the operation parameters. For example, in the 1.21 release,
there are 502 operations where the `dryRun` parameter is repeatedly
documented as operation level parameters rather than a global parameter that
can be shared and referenced. 

In KAB, such parameters are analyzed and extracted into a dedicated collection
and then referenced from the operations. For example,

```yaml
  operationId: createBatchV1NamespacedJob
  parameters:
    # 'body' is an operation-level parameter
    - name: body
      in: body
      required: true
      schema: "#/definitions/io.k8s.api.batch.v1.Job.CREATE"
    # 'dryRun' is a shared parameter
    - $ref: "#/parameters/dryRun"
    # 'namespace' is a shared parameter
    - $ref: "#/parameters/namespace"
    # 'pretty' is a shared parameter
    - $ref: "#/parameters/pretty"
```

## 3. Presenting the API specification

There are some existing documentation and tools for users to browse through
the huge collection of API groups, resources and operation definitions.

- The "k8s.io/website" has been serving the API documentation using a huge
  HTML generated from the OpenAPI specification. With all contents rendered in
  a single page, user experience is not so good.

- There is an effort to break the huge page into several groups, each of which
  generated as a Markdown text file and then rendered using the Hugo website
  framework. This effor only improves the site consistency with respect to its
  look and feel, in addition to reduced time to load the pages.

- Other OpenAPI browsers are available too. But most of those tools are
  organizing the API specification firstly by the HTTP verbs and paths.
  To Kubernetes users, the resource types (data models) are equally important
  or even more important than the operation definitions.

The KAB tool provides several features to ease the navigation of the k8s API:

- It provides a micro-service packaged into a single, self-contained docker
  image ready to run on either a server box or on a local machine.
- It provides easy navigation of all resources, definitions, operations,
  and the links among them.
- It simplifies the navigation of resources, definitions, operations across
  API versions and group versions.
- With one mouse click, you can perform comparisons between different API
  versions, group versions.
- You can inspect the change history of all resources, definitions or
  operations.
- Using the builtin YAML editor with in-place suggestions, you can easily
  compose a resource definition and then download/export it into a YAML or
  JSON for to use.

In addition to the content derived from the Kubernetes OpenAPI specification,
the KAB tool also provides documentation closely related to using the API,
such as the various methods partially updating an existing object, the HTTP
headers, the common labels, the pagination mechanism etc.

Any suggestions on feature improvments are always welcomed.
 
## 4. About future direction

Besides the limitations outlined above, the current practice of generating
Kubernetes API specification from source code may become a huge burden to
carry on for the whole community. Below are some of my thoughts about the
future development and maintenance of the API specification.

- Revising comments for documentation fixes is never the first priority for
  the developer community although we all appreciate the "API first" nature
  of a micro-service platform. There are always more urgent issues to fix,
  more compelling features to land. Pull requests (PRs) that tweak the
  documentation usually takes months to years to get reviewed and merged.

- The many details of the API definition cannot be captured in the source code
  comment in a structured manner. We will have to rely on the validation logic
  to perform error checking.

- The developers have to be cautious about code comments. For example, they
  are not supposed to freely add some to-do items as comments because such
  comments may get captured into the user-facing documentation.

- Making the source code the single source of truth had enabled the community
  to land features quickly without severely degrade the quality of
  documentation. However, as the Kubernetes core becomes more and more stable,
  we may need to rethink the relationship between API specification and the
  feature implementation.

Hopefully, this KAB tool can help raise the awareness and concerns and trigger
further discussions on improving the Kubernetes API documentation.

## 5. Technology Stack

- Python 3.6x, 3.7x (3.7.3 is used in development)
- Django 3.x (3.1.5 used in development)
- Bootstrap 4.3.x (4.3.1 is used in development)


kab/ChangeLog
CHANGES
=======

0.13.0
------

* Update build to fix setup errors
* Update build.sh
* Lock setuptools version
* Lock setup tools
* Turn off DEBUG mode
* Fix Dockerfile
* Update setup.cfg for version
* Update compiled data for 1.21-1.27
* Add 1.27 into the system, deprecate 1.20
* Update feature gates data for 1.27
* Fix typo
* Lock requirements
* Add diff between 1.26 and 1.27
* Update description
* Fix typos in 1.27
* Update operations for 1.27
* Update authentication/v1beta1 (1.27)
* Update ValidationRule (1.27)
* Remove storage/v1beta1 CSIStorageCapacity (1.27)
* Upgrade resource/v1alpha1 to resource/v1alpha2 (1.27)
* IPAddress in networking/v1alpha1 GV (1.27)
* Updates to the discovery/v1 API group (1.27)
* Updates to the core v1 API group (1.27)
* Add ClusterTrustBundle APIs (1.27)
* Update batch/v1 API group version docs (1.27)
* Update authentication/v1alpha1 and authentication/v1beta1 APIs (1.27)
* Update apps/v1 API group version (1.27)
* Updates to admissionregistration v1alpha1 API GV (1.27)
* New matchConditions for MutatingWebhook and ValidatingWebhook
* Initial data for version 1.27 (copied from 1.26)
* Diff 1.26 1.27 on 2023-03-28
* Update parameters for 1.27
* Original files for 1.27-rc.0
* Add parameter \`sendInitialEvents\`
* Update parameters for original spec

0.12.0
------

* Bump python package versions
* Update Dockerfile to use Alpine 3.17
* Fix feature gates JSON
* Update Dockerfile for 1.26 release
* Add 1.26 into const list
* Update feature gates for 1.26
* Update compiled data for 1.20-1.26 APIs
* Update settings to include 1.26
* Update generated swagger for 1.25
* Add back missing definition ClusterCIDE for 1.25
* Fix errors found in 1.26 APIs
* Update diff between 1.26 and 1.25
* Update meta API v1/status
* Add resource.k8s.io/v1alpha1 PodScheduling resource and operations (1.26)
* Add resource.k8s.io/v1alpha1 ResourceClaimTemplate resource and operations (1.26)
* Add resource.k8s.io/v1alpha1 ResourceClaim resource and operations (1.26)
* Add API resource list API for authentication/v1alpha1
* New resource.k8s.io/v1alpha1 ResourceClass type and operations (1.26)
* Update PDB for unhealthyPodEvictionPolicy field (1.26)
* Update IPv6 notation example for cluster CIDR (1.26)
* Update IngressStatus for load-balancer status. (1.26)
* Fix typo in IPBlock definition (1.26)
* Add flowcontrol/v1beta3 version resources and operations (1.26)
* New fields for borrowing and lending in LimitedPriorityLevelConfiguration (1.26)
* Remove EndpointSliceTerminatingCondition feature gate
* Drop flowcontrol/v1beta1 API group (1.26)
* Add TypedObjectReference for cross-namespace data source for PVC
* NodeTaintPolicy and NodeAffinityPolicy graduated to Beta from Alpha
* New claims field in resource requirements struct. (1.26)
* Fix typo in ReplicationControllerStatus
* Update PodSecurityContext field description. (1.26)
* Update PersistentVolumeClaimSpec field descriptions (1.26)
* Mark configSource in PodSpec as completely removed. (1.26)
* New resourceClaims and schedulingGates field in Pod spec (1.26)
* batch/v1 JobStatus.UncountedTerminatedPods GA (1.26)
* Drop autoscaling/v2beta2 API (1.26)
* Document the authentication/v1alpha1 SelfSubjectReview API (1.26)
* Document the new 'ordinals' field for StatefulSet. (1.26)
* Fix typo in ReplicaSetStatus description
* Add admissionregistration.v1alpha1 operations (1.26)
* Add admissionregistration.v1alpha1 resouce types (1.26)
* Change 1.26: RuleWithOperations extended attributes
* 1.26 API initial check in (copied from 1.25)
* Initial API for 1.26 (rc-1)
* Fix parameter \`allowWatchBookmarks\` description

0.11.0
------

* Lock requirements for 0.11.0 release
* Update dockerfile for new release
* Update index and settings for 1.25
* Update supported versions
* Update features for 1.25
* Recode changes in 1.25
* Update definitions for 1.25
* networking v1alpha1 API - ClusterCIDR
* PodFailurePolicy related definitions
* Skip meaningless changes in the spec
* Update definitions for 1.25 (1)
* 2022/08/15 version of swagger for 1.25
* Update outdated info in 1.23 and 1.24
* Resync 1.25 API changes
* Resync 1.25 API changes
* Update data for 1.25
* Clarify timeZone limite in CronJob spec
* 2022/07/08 version of swagger for 1.25
* Fix typo in autoscaling MetricTarget type
* Clarify the container resource metric requirements
* Clarify that metadata for SubjectAccessReview must be empty
* Update StatefulSet update spec
* StatefulSet spec update constraints
* Apply name constraings for webhooks
* Fix reference error in 1.24
* Fix CronJob reference issue
* Remove definitions and operations for 1.25
* Relax requirements
* Add 1.25 baseline, from 1.24

0.10.0
------

* Fix definition comparison
* Fix version list
* Prepare 0.10 release
* Update feature gates for 1.24
* Update index.json
* Update 1.24 swagger
* Cleanse 1.24 API definition
* Update 1.24 upstream API definition
* Remove some debug calls
* Update dockerfile for 1.24
* Update versions supported
* Add data for 1.24
* Update 1.24 new swagger
* Fix 1.24 nits
* Add 1.23 and 1.24 diff
* Normalized 1.24 first round
* 1.24 original spec
* Update decompose script
* Update setup.cfg for dev version 0.9.3
* Replace ADD by COPY in Dockerfile

0.9.2
-----

* Update base image again
* Upgrade dockerfile again
* Update base image to alpine3.14
* Fix django function for 4.x
* Update requirements to latest version
* Bump version to 0.9.2
* Add usage in the README

0.9.1
-----

* Fix Dockerfile
* Update setup.cfg for 0.9.1
* Update Dockerfile for using Alpine

0.9.0
-----

* Update feature gate parsing for beta1
* Update feature gates data for 1.23
* Fix Dockerfile
* Update Docker file
* Update setup.cfg
* Update swagger version number
* Add 1.23
* Add data for 1.23
* Remove watch operations in 1.23
* Improve compose script
* Fix data in 1.23
* Diff between 1.22 and 1.23
* Initial version for 1.23
* Fix 1.22 ingress status put
* Add 1.23 original

0.8.0
-----

* Update feature gate for 1.22

kab/kab/consts.py
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
# implied.
# See the License for the specific language governing permissions and
# limitations under the License.

_HTTP_SCHEMES = (
    HS_HTTP,
    HS_HTTPS,
) = (
    "HTTP",
    "HTTPS",
)

HTTP_SCHEMES = [
    (HS_HTTP, HS_HTTP),
    (HS_HTTPS, HS_HTTPS),
]

# Service Authentication Type
SAT_HTTP_BASIC = "HTTP Basic"
SAT_STATIC_BEARER_TOKEN = "Static Bearer Token"
SAT_SSL_CERTS = "SSL Certificates"

AUTH_TYPES = [
    (SAT_HTTP_BASIC, SAT_HTTP_BASIC),
    (SAT_STATIC_BEARER_TOKEN, SAT_STATIC_BEARER_TOKEN),
    (SAT_SSL_CERTS, SAT_SSL_CERTS),
]

# Authentication Request
ARC_GET = "GET"
ARC_POST = "POST"
AUTH_REQUEST_CMD = [
    (ARC_GET, ARC_GET),
    (ARC_POST, ARC_POST),
]

# HTTP DATA LOCATION
HDL_HEADER = "Header"
HDL_BODY = "Body"
HTTP_DATA_LOCATION = [
    (HDL_HEADER, HDL_HEADER),
    (HDL_BODY, HDL_BODY),
]

API_VERSIONS = [
    # ("1.13", "2018-12-04"),
    # ("1.14", "2019-03-26"),
    # ("1.15", "2019-06-20"),
    # ("1.16", "2019-09-19"),
    # ("1.17", "2019-12-10"),
    # ("1.18", "2020-03-26"),
    # ("1.19", "2020-08-27"),
    # ("1.20", "2020-12-09"),
    ("1.21", "2021-04-08"),
    ("1.22", "2021-08-04"),
    ("1.23", "2021-12-07"),
    ("1.24", "2022-05-04"),
    ("1.25", "2022-08-23"),
    ("1.26", "2022-12-13"),
    ("1.27", "2023-04-12"),
]
